// tc-source-content-updater/lib/resources/errors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var PARAMETER_ERROR = exports.PARAMETER_ERROR = "Parameter Error";
var CATALOG_CONTENT_ERROR = exports.CATALOG_CONTENT_ERROR = "Catalog Content Error";
var RESOURCE_NOT_GIVEN = exports.RESOURCE_NOT_GIVEN = 'Resource entry was not given';
var SOURCE_PATH_NOT_GIVEN = exports.SOURCE_PATH_NOT_GIVEN = 'Source path not given';
var SOURCE_PATH_NOT_EXIST = exports.SOURCE_PATH_NOT_EXIST = 'Source path does not exist';
var OUTPUT_PATH_NOT_GIVEN = exports.OUTPUT_PATH_NOT_GIVEN = 'Output path was not given';
var ERROR_PARSING_BIBLE = exports.ERROR_PARSING_BIBLE = 'Error Parsing Bible';
var MANIFEST_MISSING_BOOKS = exports.MANIFEST_MISSING_BOOKS = 'Manifest is missing the books entry';
var RESOURCES_PATH_NOT_GIVEN = exports.RESOURCES_PATH_NOT_GIVEN = 'Resources path not given';
var UNABLE_TO_CREATE_TW_GROUP_DATA = exports.UNABLE_TO_CREATE_TW_GROUP_DATA = 'Unable to create tW Group Data from Bible source';
var UNABLE_TO_MOVE_RESOURCE_INTO_RESOURCES = exports.UNABLE_TO_MOVE_RESOURCE_INTO_RESOURCES = 'Unable to move the resource files into the resources directory';
var UNABLE_TO_DOWNLOAD_RESOURCES = exports.UNABLE_TO_DOWNLOAD_RESOURCES = 'Unable to download resources';
var UNABLE_TO_UNZIP_RESOURCES = exports.UNABLE_TO_UNZIP_RESOURCES = 'Unable to unzip resources';
var FAILED_TO_PROCESS_RESOURCE = exports.FAILED_TO_PROCESS_RESOURCE = 'Failed to process resource';
var LANGUAGE_LIST_EMPTY = exports.LANGUAGE_LIST_EMPTY = 'Language list is empty';
var MANIFEST_NOT_FOUND = exports.MANIFEST_NOT_FOUND = 'Manifest not found';
var INVALID_BOOK_CODE = exports.INVALID_BOOK_CODE = 'Invalid Book Code';
